yt/frontends/cf_radial/data_structures.py

```diff
@@ -0,0 +1,269 @@
+        #   self.grid_left_edge         (N, 3) <= float64
+        #   self.grid_right_edge        (N, 3) <= float64
+        #   self.grid_dimensions        (N, 3) <= int
+        #   self.grid_particle_count    (N, 1) <= int
+        #   self.grid_levels            (N, 1) <= int
+        #   self.grids                  (N, 1) <= grid objects
+        #   self.max_level = self.grid_levels.max()
+        self.grid_left_edge[0][:] = self.ds.domain_left_edge[:]
+        self.grid_right_edge[0][:] = self.ds.domain_right_edge[:]
+        self.grid_dimensions[0][:] = self.ds.domain_dimensions[:]
+        self.grids = np.empty(self.num_grids, dtype="object")
+        for i in range(self.num_grids):
+            g = self.grid(i, self, self.grid_levels.flat[i], self.grid_dimensions[i])
+            g._prepare_grid()
+            g._setup_dx()
+            self.grids[i] = g
+        self.max_level = 1
+        pass
+
+    def _populate_grid_objects(self):
+        # For each grid, this must call:
+        #   grid._prepare_grid()
+        #   grid._setup_dx()
+        # This must also set:
+        #   grid.Children <= list of child grids
+        #   grid.Parent   <= parent grid
+        # This is handled by the frontend because often the children must be
+        # identified.
+        pass
+
+
+class CFRadialDataset(Dataset):
+    _index_class = CFRadialHierarchy
+    _field_info_class = CFRadialFieldInfo
+
+    def __init__(
+        self,
+        filename,
+        dataset_type="cf_radial",
+        grid_shape=(40, 200, 200),
+        grid_limits=((0.0, 2000.0), (-1e5, 1e5), (-1e5, 1e5)),
+        storage_filename=None,
+        units_override=None,
+    ):
+        self.fluid_types += ("cf_radial",)
+        self._handle = xr.open_dataset(filename)
+        self.refine_by = 2
+        new_filename = filename[:-3] + "_grid.nc"
+        if "x" not in self._handle.coords:
+            if not os.path.isfile(new_filename):
+                from yt.utilities.on_demand_imports import _pyart
+
+                pyart = _pyart.pyart
+
+                radar = pyart.io.read_cfradial(filename)
+                self.grid_shape = grid_shape
+                self.grid_limits = grid_limits
+                grid = pyart.map.grid_from_radars(
+                    (radar,), grid_shape=self.grid_shape, grid_limits=self.grid_limits
+                )
+                mylog.warning(
+                    "Saving a cartesian grid for file %s at %s. Data will be loaded "
+                    "from the cartesian grid.",
+                    filename,
+                    new_filename,
+                )
+                grid.write(new_filename)
+            self._handle = xr.open_dataset(new_filename)
+            filename = new_filename
+        super().__init__(filename, dataset_type, units_override=units_override)
+        self.storage_filename = storage_filename
+        # refinement factor between a grid and its subgrid
+        # self.refine_by = 2
```

Comment:
Even if unused, this should be set to avoid breaking other methods (this attribute doesn't have a default value)
```suggestion
        self.refine_by = 2
```

Reply:
this was set higher up... deleted that line, uncommented this one.